hpct-ldap-server/src/servers/ldap.py

```diff
@@ -0,0 +1,503 @@
+            stderr=subprocess.STDOUT,
+        )
+        if rc != 0:
+            raise Exception(f"Unable to set debconf for slapd.")
+
+    def tls_gen(self, org=None) -> None:
+        """Create CA cert.
+
+        Parameters
+        ----------
+        org : str
+            Organization name.
+        """
+        if org:
+            # Create Private Key for Certificate Authority(CA)
+            pkc_args = [
+                "certtool",
+                "--generate-privkey",
+                "--bits",
+                "4096",
+                "--outfile",
+                "/etc/ssl/private/mycakey.pem",
+            ]
+            rc = subprocess.call(pkc_args, stdout=subprocess.DEVNULL, stderr=subprocess.STDOUT)
+
+            if rc != 0:
+                raise Exception("Unable to create private key.")
+
+            # CA Template
+            ca_template = [f"cn = {org}", "ca", "cert_signing_key", "expiration_days = 3650"]
+            with open("/etc/ssl/ca.info", "w") as f:
+                for line in ca_template:
+                    f.write(f"{line}\n")
+
+            # Create CA Certificate
+            ca_cert_args = [
+                "certtool",
+                "--generate-self-signed",
+                "--load-privkey",
+                "/etc/ssl/private/mycakey.pem",
+                "--template",
+                "/etc/ssl/ca.info",
+                "--outfile",
+                "/usr/local/share/ca-certificates/mycacert.crt",
+            ]
+
+            rc = subprocess.call(ca_cert_args, stdout=subprocess.DEVNULL, stderr=subprocess.STDOUT)
+
+            if rc != 0:
+                raise Exception("Unable to create certificate authority.")
+
+            # Add CA certificate to trusted certs
+            rc = subprocess.call(
+                ["update-ca-certificates"], stdout=subprocess.DEVNULL, stderr=subprocess.STDOUT
+            )
+
+            if rc != 0:
+                raise Exception("Unable to add CA certificate to trusted certs")
+
+            # Get Server Hostname
+            result = subprocess.run(["cat", "/etc/hostname"], capture_output=True, text=True)
+
+            # Private Key for Server
+            pks_args = [
+                "certtool",
+                "--generate-privkey",
+                "--bits",
+                "2048",
+                "--outfile",
+                "/etc/ldap/ldap01_slapd_key.pem",
+            ]
+            rc = subprocess.call(pks_args, stdout=subprocess.DEVNULL, stderr=subprocess.STDOUT)
+
+            if rc != 0:
+                raise Exception("Unable to create private key for server.")
+
+            # Template for Server Certificate
+            sc_template = [
+                f"organization = {org}",
+                f"cn = {result.stdout}",
+                "tls_www_server",
+                "encryption_key",
+                "signing_key",
+                "expiration_days = 365",
+            ]
+            with open("/etc/ssl/ldap01.info", "w") as f:
```

Comment:
Is `ldap001.info` a standard name?

Reply:
No, it's supposed to match the server hostname. Will replace with the juju hostname ie "juju-abc123-0".